Multiple tasks

bitterdev opened the pull request:
This pull request contains the following tasks:

- Bootstrap 4 dropdowns in front-end require two clicks to activate (#8938 on GitHub)
- Sign in as user does not update the dashboard navigation pane (#8830 on GitHub)
- Add "Calendar" to Permissions (#8486 on GitHub)
- User workflows - User activation does not trigger on admin email validations #8416 (#8416 on GitHub)
- Big clipboard hangs editing UX (#7655 on GitHub)
- Orphaned Blocks Option in Add Block Panel (#3234 on GitHub)
- Add configuration for email logging (#8768 on GitHub)
- Add logging for file uploads and file deletions (#8979 on GitHub)

Comments:
- For task #3234 you need to update the bedrock repo and rebuild the assets.
- To fix the bootstrap issue i needed to declare bootstrap as an webpack external and load it by the concrete5 asset manager (same like jQuery, Vue...). I need to do the same with moment-library because otherwise the calendar block was not working. There is some inline js which requires moment and i think there are a lot of other places too in the core which requires moment. Im not happy with adding so many fundamental dependencies because it stops sites from beeing lightweight. But thats the disadvantages of webpack. Just adding moment to the asset groups like fullcalendar like in previous core versions was not working to because all occurrences of requireAsset('fullcalendar') etc. was already removed from the core. So i needed to add them globally.
- I needed to create a separate table called "UsedAreas" to track all rendered areas with the associated theme, page type, page id... Otherwise there is not way to detect the orphaned blocks.
- Navigation cache in theory is a nice feature but as you can see in my pull requests there are a lot of occurrences where it is required to clear the the cache. Otherwise you will get outdated data and have issues like #8830. Maybe we need to implement some other concepts to improve this in the future like tracking important events and auto clear the cache on affected changes (even better just update the changes within the cache)
- I don't checked this in detail but most libraries required localizations. While declaring momentjs as an external i have implemented the translations for too. I have used the same code like in previous concrete5 versions. So not a big deal. But what is with any new javascript components that are part of the web pack bundles. Are they localized too? Just some thoughts. This should be checked...
- All in all i'm not really happy with the web pack / laravel mix solution. I see too many disadvantages... The current asset manager is old but gold. I know the asset manager was bad for SEO (to many javascript + css) but now we are running into the same issues when declaring so many externals. And if we don't do so we get unexpected behaviours, issues because of missing references within inline js etc. Extending the old asset manager and implementing some kind of a public asset registry (maybe with yarn behind) and some better bundle technics were the right way IMO. But this is just my opinion.
- Issue #6440 (wrong presentation of all-day-events) i was not able to reproduce. On my local environment everything will be display correctly. Maybe this issue was already fixed in the past? (see pictures attached)
![Bildschirmfoto 2020-08-21 um 14 01 00](https://user-images.githubusercontent.com/15018219/90902318-90c6c280-e3cc-11ea-87a6-9c15cd9d968d.png)
![Bildschirmfoto 2020-08-21 um 14 00 52](https://user-images.githubusercontent.com/15018219/90902320-915f5900-e3cc-11ea-9c3a-984d9e308098.png)
![Bildschirmfoto 2020-08-21 um 14 01 11](https://user-images.githubusercontent.com/15018219/90902309-8efcff00-e3cc-11ea-9d05-6eb798231360.png)
- Task #8501 (uploading => duplicate files) i was not able to fix within this sprint because at the start of the week the task #8754 was not ready which was required for testing. Will continue this task by the next week.

bitterdev commented:
Added 
- Customize Design Dashboard not working (#8787 on GitHub)
- Mail logging fails with large attachments (#5473 on GitHub)
- V9: Adding or editing feature block thows error (#8488 on GitHub)

aembler commented:
Hey Fabian, I've finally had a chance to go through all of these. Looking really good! Here are some general notes (and after that I've got some some things that I think will be pretty easy for you to address):

## Notes

* Your bedrock modifications look good! The way you're structuring your CSS and JS changes look good to me.
* Regarding bootstrap: that makes total sense to me. It's not ideal but like you said we're probably going to have to do that with some of these things. I want to use bedrock to make it so that we can have fewer JS and CSS files everywhere - but if we have to have assets defined for things like Vue, moment and Bootstrap (especially for things like these external dependencies) then it's fine that they're external.
* Mail logging, feature block changes, file uploading logging and other miscellaneous changes look good. 

My real concern really is only in the Orphaned Blocks functionality. 

## Changes

* I'm not sure we need the UsedAreas database table, and I'm a little concerned about attempting to insert into it every time we render something on the page. I think instead we should be looking to the `CollectionVersionBlocks` table, which joins Pages (cID) on a particular version of the page (cvID) to Areas (arHandle) and block IDs (bID). I think a typical flow would look like this:

1. User loads edit mode and clicks on the plus icon to load the add panel.
2. Add Panel Loads.
3. A new routine looks at the current page template being loaded, and determines - in real time, maybe by inspecting the template, or through some pre-cached routine somewhere, what areas are actively in use in the page template. This would be done by maybe doing a preg_match() within the PHP template against the Area() class? Maybe something smart could be done using reflection on the template itself? This would grab a list of all the areas that are actively defined within the page templates.
4. Then, a routine against the current version (recent) of the page would be run to grab all blocks in the `CollectionVersionBlocks` table against this version of the page. Orphaned blocks would be determined if any of the blocks in the `CollectionVersionBlocks` database table are in areas that aren't found in the routine that looks at the actual page template.

With all this in mind, can you do the following?

1. Can you remove the UsedAreas table and change the `getOrphanedBlocks` method to work in this way? 
2. Can you add the new routine that scans the page templates in real time and determines what areas are used in those templates?
3. Can you move "Orphaned Blocks" to the bottom of the dropdown menu, and only show it if there are actually orphaned blocks (blocks in the `CollectionVersionBlocks` table that are in areas that aren't present on the page?)
4. Make the Orphaned blocks panel actually render like the clipboard? I saw your earlier post about having difficulty doing this because not knowing what areas the blocks were in - but I think you should be able to just pass them to whatever area they're in based on what's in the `CollectionVersionBlocks` row for that particular block. Does that make sense?

Thanks! This is going to be a really nice feature. And I won't let this pull request sit so long before I review it again. Sorry about that - we had a big launch last week. We still have to support it and work on it going forward, but I won't be so focused on making sure that goes smoothly. I'll be able to devote more time to review.

bitterdev commented:
I know about the CollectionVersionsBlocks-Table and it’s already used for the query to determinate the orphaned blocks.

See src/Page/Collection/Collection.php, line 996.

The reason for why I have created this table is to get the current areas for the selected page in combination with the theme, version etc.

preg_match() is not a good solution because this needs to be work with all themes and you don’t can cover up all possibilities by all theme developers, see:

```
new   Area('Test');

new Area(
  'Name'
)

new GlobalArea(            'Name');
```

If you really want to cover all cases you need a lexer/parser engine.

I have implemented such a [library](https://github.com/nikic/PHP-Parser) to my Entity Designer add-on and it works like a charm. It will return you a tree which you can parse. The problem is that this library is very overloaded for such a small use-case.

There is also a less comfortable but more lower level native method to parse a php file, called „token_get_all“, see https://www.php.net/manual/de/function.token-get-all.php.

With this you could parse a file, strip all whitespaces and check for Area|GlobalArea, (, String, ). And collect the strings.

But what is when somebody is using something like this:

`new SomeCustomOverrideArea('test')`

This makes it impossible to analyse the code further.

So you need to decide if you prefer the code parser solution anyway with keeping in mind that custom extended areas won't work with the orphaned block feature. But if that's okay it's an nicer approach then storing everytime the used areas in a table.

Because of the rendering in clipboard layout: It was long time ago that i have worked on that task but i remember that there were big issues while rendering the scrapbook layout when the destination area is no longer available. So this will definitely take more work to adapt the code. Can i have 4hrs more for this task? I think that should be enough for rendering the orphaned blocks in the scrapbook layout.

bitterdev commented:
Implemented Task Form File/Image Upload Duplicate Files #8501

bitterdev commented:
@aembler I have rebased my repository before i have commited the last changes but now i'm getting errors because of another migration.
```
https://travis-ci.org/github/concrete5/concrete5/jobs/725193043
1) Concrete\Tests\Update\NewMigrationsTest::testNewMigrations
There shouldn't be any migration with an ID lower than 20200903201537
Failed asserting that two arrays are identical.
--- Expected
+++ Actual
@@ @@
-Array &0 ()
+Array &0 (
+    0 => '20200818000000'
+    1 => '20200818000001'
+    2 => '20200818000002'
+    3 => '20200818000003'
+)
```

Is this issue up to me or is this an existing issue from the dev branch?

bitterdev commented:
Next issue when using code detection... Most themes including files by $this->inc() method. And each included file can contain multiple areas too. It is very hard to find all depended files.
One approach could be to scan all php files within a theme ditectory.
But what is e.g. when a view of the elements Directory is requested or if a Developer dont follow the best practices of c5 develolment and using include or require statements etc.? As you can see it is really hard to scan the code on the fly.

bitterdev commented:
And if you scan all files you dont know which areas belongs to which template.

Another option is to extend the theme base Controller with methods like getAreasByTemplate($templateName). But this would it make it even harder to develop themes. Most theme developers are designers with less skills about the architecture of c5.

So currently i think the UsedAreas table is ugly yes but it seems to be the most stable solution for determing the available areas for a theme/template...

JohntheFish commented:
> 3\. A new routine looks at the current page template being loaded, and determines - in real time, maybe by inspecting the template, or through some pre-cached routine somewhere, what areas are actively in use in the page template. This would be done by maybe doing a preg_match() within the PHP template against the Area() class? Maybe something smart could be done using reflection on the template itself? This would grab a list of all the areas that are actively defined within the page templates.

Whilst a clean solution, this will break on themes that determine area names dynamically, such as using page attributes or path to determine areas names or even areas to show.

aembler commented:
That's true – I imagine those sites will not be able to use the Orphaned Blocks feature.

JohntheFish commented:
As long as the orphan determination generates false positives rather than misses anything, I suppose a manual decision can be made. 

Once beyond development, the main source of orphaned blocks is layouts that have been removed. Is there anything that can be done specific to layouts that would be foolproof?

JohntheFish commented:
Have you looked at how the Areas Cleanup addon in the marketplace identified zombie areas and hence orphaned blocks? 
https://www.concrete5.org/marketplace/addons/areas-cleanup/

bitterdev commented:
Hi John :) as far as i can see this addon is listing all availble areas/blocks. This is no Problem. The problem currently is to determ the unused areas. Currently i have implemented a Tracking table but thats not the better solution.

JohntheFish commented:
Hmm yes. I just remembered from the PRB that it was attacking a similar problem. Obviously not in a way that you could use.

Simply scanning a source file with a regex or token extractor will have way too many holes in it to be useful.
- Themes usually structure templates through including elements, so it would immediately need to scan included elements. 
- Any area name assigned through code would be lost $a = new Area($variable_name)
- Pseudo areas within layouts never appear as a name in theme templates.

My gut says 'there must be a way of working this out from current data/tables', but I can't say what that is. I suspect Andrew's objection to an extra table comes from a similar gut feeling that this should be achievable with current data.

JohntheFish commented:
It may be that a combination of algorithms are needed. For example, blocks orphaned by removing a layout could be identified by a completely different algorithm to blocks orphaned by a theme or template change.

aembler commented:
Hey @bitterdev - Yes, you can definitely have 4 more hours to finish up the task. I like the idea of using the `token_get_all` function from PHP. I think we can standardize on the `Area` and `GlobalArea` classes.

Will this catch everything? No - any custom subclasses or custom areas won't be caught. But it's still a lot better than what we're doing today, which is nothing. Maybe we could make the things that we scan for an array that we could configure? Stored in the app.php config file? That way if there were custom area subclasses that a developer was using that they wanted to work with this functionality, they could add/define them there?

bitterdev commented:
Yes thats fine with the extendable array in a config file. This allows developers to extend the classes. I would suggest to scan the areas from the page template and all global areas from all php files inside the theme directory? But anyway this will not give you all areas... For example if you have included areas outside the page template (e.g. in the view folder or in another element of the theme). So this solution has not 100% coverage.

aembler commented:
Yeah, I think if we can scan all the files within the theme directory that will be enough. It's not perfect but it's pretty good... it's also a good reason as to why people might consider using the elements/ directory within a theme instead of within the application directory.

JohntheFish commented:
How does this handle layouts that have been deleted?

aembler commented:
Deleting layouts delete any blocks that are inside them.

JohntheFish commented:
Is that new functionality? Last time I looked at the database after deleting a layout the content blocks within it still existed in the Blocks and btContentLocal tables. That would have been one or two core versions ago.

bitterdev commented:
@JohntheFish i dont have checked this. A question for andrew. @aembler there is another important comment from John in the comments above that you should read before i start further development... Dynamic area names will also not work with the code parsing solution. Is that okay for you?

JohntheFish commented:
Is the content of the orphan blocks panel worked out before or after a page is rendered? Or is it worked out asynchronously by ajax when the panel is opened?

bitterdev commented:
Currently all block related panels are loaded by ajax (after clicking on open panel or switching the panel view)

bitterdev commented:
John ur the man! Just your comment let me think into the right direction.. Because its ajax the areas can be temporary stored into a global var and passed to the query params of the panel url. Then we have all used areas and can detect the orphaned blocks without having a temporary table or analyzing the code

JohntheFish commented:
Being naive about the details behind this, if I were starting on the problem from scratch, the kind of algorithm I would be looking at is:

If a user has relevant permissions and a page is in edit mode
- note the blocks and areas rendered as the page loads
- save the list somewhere temporary, such as in the user's session or appended as json hidden in the page footer

When the panel is opened (and I guess this is what you have already solved)
- refer to the list recently noted for blocks and areas actually in use

JohntheFish commented:
crossed posts

bitterdev commented:
I have changed the logic, removed the table and using a session variable to temporary store the areas. Works like a charm. I will add the design update on monday.

mlocati commented:
Just a warning: I didn't check the code, but we should be careful if we use the session to store this data. What of a user is editing 2 pages in 2 different tabs? When opening the panel of one of page 1 do we risk to actually use the data of page 2?

bitterdev commented:
True. This case isnt covered by this solution. Will change the code to use a global var or json in the footer like John suggested and pass the params to the query of the panel url.

JohntheFish commented:
It could stay in the session if indexed by cID, with some garbage collection to clean up the session and stop it becoming overwhelmed with orphan data as pages are edited.
Not saying doing so is a best solution. Just presenting an alternative.

bitterdev commented:
Kind of garbage collector is already implemented. The areas are only tracked to the session in edit mode and cleared when leaving the edit mode. Just adding the cID is definitly less effort and bring same results. So i will choose this solution.

bitterdev commented:
@aembler i have changed the logic, remove the session object and retrieving the area handles from the DOM instead and passing them to the query of the panal urls. This is is relevant peace of code for determing the used blocks:
```
var areaHandles = [];
$(".ccm-area").each(function() {
  areaHandles.push($(this).data("areaHandle"));
});
```
Furthermore i have changed the view to look like the clipboard panel. To test this you need to update bedrock repo first.

It was required to rebase to be able to merge (there was a lot of conflicts). 

Unfortunately now 2 automatic tests failed:
1) In [this commit](https://github.com/concrete5/concrete5/commit/a0429d99597de98222ce298eac12cb78c467b836) the constructor of the CsvWriter has been changed which leads into 2 automatic test failures.
2) There is also another issue with testNewMigrations.

aembler commented:
Thanks @bitterdev .

I'm manually merging this in with another pull request (because we had to let it sit so long it's really out of date, which is why you're seeing the test phpunit errors, as well as issues building the assets). https://github.com/concrete5/concrete5/pull/9076

Once that passes tests I'm going to merge this and close this and update all the linked tickets. In general things look really good here. I really like the solution that you and @JohntheFish came to regarding the areas. Very clever, but also very simple (which is the best kind of solution!)

I still have some concerns about the orphaned blocks functionality, which is why I'm temporarily disabling that functionality. (See: https://github.com/concrete5/concrete5/pull/9076/commits/a3ceb718000ac8a68aab69a3b19d1f6af3038abe#diff-bf126826b8eca421801edc03e59fb652R117) Here are my concerns about the orphaned blocks.

1. When you add blocks from this it reloads the page. This doesn’t happen whenever we add anything else from this panel. We don’t want a full page reload. I know you're aware of this because it was in your comment earlier, I just think we need to have time to fix this.
2. In the original pull request, if you had no orphaned areas, it shows all the blocks in the sidebar. I think this was just a logic error and I fixed it (See: https://github.com/concrete5/concrete5/pull/9076/commits/46096d8e9a9a7ab4124817c735aa74ab1ce77f19#diff-bf126826b8eca421801edc03e59fb652R78)
3. We need to modify my new `showOrphanedBlocks` method to detect whether the option should be shown. If that method returns true we show the option in the dropdown, otherwise not.
4. If you drag an orphaned block into the page not at the top of the area, right now it reloads and the block is at the top of the area. It doesn’t respect you adding the blocks into the middle of the area after other blocks. Probably related to number 1.
5. I wasn’t able to delete an orphaned block from the panel, I think the z-index of the delete link needs to be higher? Or maybe something else is happening?

At any rate, this has been sitting too long and shouldn't be held up. I think in the future we should break these out into smaller pull requests. I know I was the one who told you it was fine to combine them, but these smaller items could have been merged weeks ago. I think smaller pull request are better. I'm going to open a new issue for these remaining things, but it's low priority. You've done great work here, thank you!

aembler commented on concrete/src/Express/Entry/Manager.php:
```diff
@@ -122,19 +123,18 @@ public function saveEntryAttributesForm(Form $form, Entry $entry)
-                $attributeValue->setAttributeKey($attributeKey);
-                $attributeValue->setEntry($entry);
-                $attributeValue->setAttributeValueObject($genericValue);
+                $akHandle = $attributeKey->getAttributeKeyHandle();
```
Hey, I'm not sure what this code is doing in the orphaned blocks commit. I totally missed it the first time. Is this an attempt to fix the duplicate attribute issue? I'm not sure the logic behind these changes but it looks pretty wrong to me. I'm reverting this.
